src/download.rs

```diff
@@ -349,9 +303,90 @@ pub fn download_to_tempfile(url: &str) -> Result<File> {
     Ok(f)
 }
 
-/// Format a size in bytes.
-fn format_bytes(count: u64) -> String {
-    Byte::from_bytes(count.into())
-        .get_appropriate_unit(true)
-        .format(1)
+struct ProgressReader<'a, R: Read> {
+    source: R,
+    length: Option<u64>,
```

Comment:
Coming back from the rest of the code, you may consider making `length` an `Option<NonZeroU64>` (and storing the formatted label too).